File: backends/js/src/types/apiResponse.ts
import { ErrorMessage } from "./errors";
import { NonceResponse, LoginResponse } from "./routes";

type Result = ErrorMessage | NonceResponse | LoginResponse;

export class ApiResponse {
  success: boolean;
  result: Result | undefined;
  constructor(success: boolean = false, result?: Result) {
    this.result = result;
    this.success = success;
  }
}

File: backends/rust/src/structs/api_response.rs
use serde::Serialize;

#[derive(Serialize)]
pub struct ApiResponse<T> {
    success: bool,
    result: T,
}

impl<T> ApiResponse<T> {
    pub fn new(success: bool, data: T) -> ApiResponse<T> {
        Self {
            success,
            result: data,
        }
    }
}
